src/signTx.c

```diff
@@ -0,0 +1,648 @@
+
+static void signTx_handleInit_ui_runStep();
+
+enum {
+	HANDLE_INIT_STEP_CONFIRM = 100,
+	HANDLE_INIT_STEP_RESPOND,
+	HANDLE_INIT_STEP_INVALID,
+} ;
+
+static void signTx_handleInitAPDU(uint8_t p2, uint8_t* wireDataBuffer, size_t wireDataSize)
+{
+	TRACE();
+
+	CHECK_STAGE(SIGN_STAGE_NONE);
+	VALIDATE(p2 == 0, ERR_INVALID_REQUEST_PARAMETERS);
+
+	txHashBuilder_init(&ctx->txHashBuilder);
+
+	ctx->currentInput = 0;
+	ctx->currentOutput = 0;
+	ctx->currentWitness = 0;
+
+	ctx->sumAmountInputs = 0;
+	ctx->sumAmountOutputs = 0;
+
+
+	struct {
+		uint8_t numInputs[4];
+		uint8_t numOutputs[4];
+	}* wireHeader = (void*) wireDataBuffer;
+
+	VALIDATE(SIZEOF(*wireHeader) == wireDataSize, ERR_INVALID_DATA);
+
+	ctx->numInputs    = u4be_read(wireHeader->numInputs);
+	ctx->numOutputs   = u4be_read(wireHeader->numOutputs);
+
+	VALIDATE(ctx->numInputs < SIGN_MAX_INPUTS, ERR_INVALID_DATA);
+	VALIDATE(ctx->numOutputs < SIGN_MAX_OUTPUTS, ERR_INVALID_DATA);
+
+	// Note(ppershing): current code design assumes at least one input
+	// and at least one output. If this has to be relaxed, stage switching
+	// logic needs to be re-visited
+	VALIDATE(ctx->numInputs > 0, ERR_INVALID_DATA);
+	VALIDATE(ctx->numOutputs > 0, ERR_INVALID_DATA);
+
+	// Note(ppershing): right now we assume that we want to sign whole transaction
+	ctx->numWitnesses = ctx->numInputs;
+
+	// Note(ppershing): do not allow more witnesses than inputs. This
+	// tries to lessen potential pubkey privacy leaks because
+	// in WITNESS stage we do not verify whether the witness belongs
+	// to a given utxo.
+	VALIDATE(ctx->numWitnesses <= ctx->numInputs, ERR_INVALID_DATA);
+
+	security_policy_t policy = policyForSignTxInit();
+	switch (policy) {
+	case POLICY_PROMPT_BEFORE_RESPONSE: {
+		ctx->ui_step = HANDLE_INIT_STEP_CONFIRM;
+		break;
+	}
+	case POLICY_ALLOW: {
+		ctx->ui_step = HANDLE_INIT_STEP_RESPOND;
+		break;
+	}
+	default:
+		THROW(ERR_NOT_IMPLEMENTED);
+	}
+	signTx_handleInit_ui_runStep();
+}
+
+// Simple for now
+static void signTx_handleInit_ui_runStep()
+{
+	ui_callback_fn_t* this_fn = signTx_handleInit_ui_runStep;
+	int nextStep = HANDLE_INIT_STEP_INVALID;
+
+	switch(ctx->ui_step) {
+	case HANDLE_INIT_STEP_CONFIRM: {
+		TRACE();
+		ui_displayConfirm(
+		        "Start new",
+		        "transaction?",
+		        this_fn,
+		        respond_with_user_reject
+		);
+		nextStep = HANDLE_INIT_STEP_RESPOND;
+		break;
+	}
+	case HANDLE_INIT_STEP_RESPOND: {
+		TRACE();
+		// advance state to inputs
+		txHashBuilder_enterInputs(&ctx->txHashBuilder);
+		ctx->stage = SIGN_STAGE_INPUTS;
+
+		// respond
+		io_send_buf(SUCCESS, NULL, 0);
+		ui_displayBusy(); // needs to happen after I/O
+		nextStep = HANDLE_INIT_STEP_INVALID;
+		break;
+	}
+	default:
+		ASSERT(false);
+	}
+	ctx->ui_step = nextStep;
+}
+
+
+static void amountSum_incrementBy(uint64_t* sumPtr, uint64_t amount)
+{
+	ASSERT(*sumPtr <= LOVELACE_MAX_SUPPLY);
+	ASSERT(*sumPtr <= LOVELACE_MAX_SUPPLY);
+	if (*sumPtr + amount > LOVELACE_MAX_SUPPLY) {
```

Comment:
Use assert?

Reply:
Nope. Assert is internal error, but this is an expected overflow, especially if an attacker feeds bad transaction.  But I guess I should use VALIDATE instead ...